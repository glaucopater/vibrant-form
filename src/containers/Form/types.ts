

export type DataType = {
    name: string;
    value?: string | number;
    isRequired?: boolean;
}

export type FormFieldType = Pick<DataType, "name" | "value">;


export type FormStateType = { [name: string]: FormFieldType };


export type FormPropsType = {
    fieldsData?: DataType[];
    action?: string;
    method?: string;
}

export type validationErrorType = {
    name: string;
    message: string;
}

export type IValidationErrors = {
    errors?: validationErrorType[];
}

export type InputPropsType = DataType & IValidationErrors & {
    onChange: (e: any, name: DataType["name"], isRequired: DataType["isRequired"]) => void;
}
