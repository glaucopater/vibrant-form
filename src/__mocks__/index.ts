import { DataType } from "../types";

export const mockFields: DataType[] = [
    {
        name: "Full Name",
        value: "John Doe",
        type: "text",
        isRequired: true
    },
    {
        name: "role",
        value: "Designer",
        type: "text",
        isRequired: true
    },
    {
        name: "department",
        value: "wooga.dev",
        type: "text",
        pattern: "wooga.",
        isRequired: true
    },
    {
        name: "salary",
        value: 900,
        type: "text",
        minValue: 1000,
        isRequired: true
    },
];