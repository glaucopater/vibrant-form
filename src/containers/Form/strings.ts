export const strings = {
    "submit": "Submit",
    "fieldIsRequired": "field is required"
}