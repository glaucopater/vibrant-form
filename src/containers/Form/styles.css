:root {
  --text-shadows: 2px 2px 4px rgba(0, 0, 0, 0.2);
  --box-shadow: 20px 20px 40px -6px rgba(0, 0, 0, 0.2);
  --custom-border-color: 1px solid rgba(255, 255, 255, 0.3);
  --background-color: rgba(255, 255, 255, 0.3);
  --white: #f1f1f1;
}


.vibrantForm {
  background: var(--background-color);
  padding: 3em;
  border-radius: 20px;
  border-left: 1px solid var(--background-color);
  border-top: 1px solid var(--background-color);
  -webkit-backdrop-filter: blur(10px);
  backdrop-filter: blur(10px);
  box-shadow: var(--box-shadow);
  transition: all 0.2s ease-in-out;
  display: flex;
  flex-direction: column;
  place-items: center;
}


@media only screen and (max-width: 600px) {
  .vibrantForm {
    padding: 1em;
  }
}

.vibrantForm p {
  font-weight: 500;
  color: var(--white);
  opacity: 0.7;
  font-size: 1.4rem;
  margin-top: 0;
  margin-bottom: 60px;
  text-shadow: var(--text-shadow);
}