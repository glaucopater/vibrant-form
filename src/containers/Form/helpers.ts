import { FormFieldType } from "./types";
import { strings } from "./strings";

export const validateFormField = (field: FormFieldType) => {
    if (!field.value) {
        return {
            message: strings.fieldIsRequired
        }
    }
}

