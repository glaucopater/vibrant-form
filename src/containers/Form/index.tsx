import React, { FormEvent } from "react";
import Submit from "../../components/Submit";
import { getClassNameProps, isNullOrEmpty, transformDataIntoFormField, validateForm } from "../../helpers";
import { FormPropsType, TransformedDataType, FormOnChangeEventType, ValidationErrorType } from "../../types";
import InputFields from "../InputFields";
import { dictionary } from "../../dictionary";
import "./styles.css";


const Form: React.FC<FormPropsType> = ({ fieldsData, action, method, withTheme, component: CustomInput }) => {
  const initialFormData = transformDataIntoFormField(fieldsData);
  const [formData, updateFormData] = React.useState<TransformedDataType>(initialFormData);
  const [formErrors, setFormErrors] = React.useState<ValidationErrorType[]>();
  const [isSubmitting, setIsSubmitting] = React.useState<boolean>();

  const sendData = () => {
    if (isNullOrEmpty(action))
      setTimeout(() => {
        setIsSubmitting(false);
      }, 3000);
    else {
      const requestOptions = {
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(formData)
      };
      if (action)
        fetch(action, requestOptions)
          .then(response => console.log(dictionary.dataSent, response))
          .catch(error => console.log('Server error: ', error))
          .finally(() => setIsSubmitting(false))
    }
  }

  const handleOnSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const errs = validateForm(formData, fieldsData);
    setFormErrors(errs);
    if (errs.length === 0) {
      setIsSubmitting(true);
      sendData();
    }
  }

  const handleOnChange = (e: FormOnChangeEventType) => {
    updateFormData({
      ...formData,
      [e.target.name]: e.target.value
    });
  }

  const formProps = {
    action, method,
    onSubmit: handleOnSubmit,
    onChange: handleOnChange
  }

  return (
    <section
      {...getClassNameProps("vibrantFormContainer", withTheme)}>
      <form
        {...getClassNameProps("vibrantForm", withTheme)}
        {...formProps}>
        <InputFields
          fieldsData={fieldsData}
          formErrors={formErrors}
          withTheme={withTheme}
          component={CustomInput}
        />
        {isSubmitting === false && <span>{dictionary.dataSent}</span>}
        <Submit isSubmitting={isSubmitting} withTheme={withTheme} />
      </form>
    </section >
  );
}

export default Form;