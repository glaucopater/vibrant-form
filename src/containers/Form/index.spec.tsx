import Form from ".";
import { mockFields } from "../../__mocks__/index";

describe("VibrantForm", () => {
  it("should render the empty component", () => {
    expect(Form).toMatchSnapshot();
  })

  it("should render the component with mock data", () => {
    const wrapper = <Form fieldsData={mockFields} />;
    expect(wrapper).toMatchSnapshot();
  })
})
