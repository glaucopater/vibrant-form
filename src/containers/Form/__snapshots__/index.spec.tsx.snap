// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VibrantForm should render the component with mock data 1`] = `
<Form
  fieldsData={
    Array [
      Object {
        "isRequired": true,
        "name": "Full Name",
        "type": "text",
        "value": "John Doe",
      },
      Object {
        "name": "role",
        "type": "text",
        "value": "Designer",
      },
      Object {
        "isRequired": true,
        "name": "department",
        "pattern": "wooga.",
        "type": "text",
        "value": "wooga.dev",
      },
      Object {
        "isRequired": true,
        "minValue": 1000,
        "name": "salary",
        "type": "text",
        "value": 900,
      },
    ]
  }
/>
`;

exports[`VibrantForm should render the empty component 1`] = `[Function]`;
